lib/experimental/full-page-client-side-navigation.php

```diff
@@ -0,0 +1,60 @@
+<?php
+/**
+ * Registers full page client-side navigation option using the Interactivity API and adds the necessary directives.
+ */
+
+/**
+ * Enqueue the interactivity router script.
+ */
+function _gutenberg_enqueue_interactivity_router() {
+	// Set the navigation mode to full page client-side navigation.
+	wp_interactivity_config( 'core/router', array( 'navigationMode' => 'fullPage' ) );
+	wp_enqueue_script_module( '@wordpress/interactivity-router' );
+}
+
+add_action( 'wp_enqueue_scripts', '_gutenberg_enqueue_interactivity_router' );
+
+/**
+ * Set enhancedPagination attribute for query loop when the experiment is enabled.
+ *
+ * @param array $parsed_block The parsed block.
+ *
+ * @return array The same parsed block with the modified attribute.
+ */
+function _gutenberg_add_enhanced_pagination_to_query_block( $parsed_block ) {
+	if ( 'core/query' !== $parsed_block['blockName'] ) {
+		return $parsed_block;
+	}
+
+	$parsed_block['attrs']['enhancedPagination'] = true;
+	return $parsed_block;
+}
+
+add_filter( 'render_block_data', '_gutenberg_add_enhanced_pagination_to_query_block' );
+
+/**
+ * Add directives to all links.
+ *
+ * Note: This should probably be done per site, not by default when this option is enabled.
+ *
+ * @param array $content The block content.
+ *
+ * @return array The same block content with the directives needed.
+ */
+function _gutenberg_add_client_side_navigation_directives( $content ) {
+	$p = new WP_HTML_Tag_Processor( $content );
+	while ( $p->next_tag( array( 'tag_name' => 'a' ) ) ) {
+		if ( empty( $p->get_attribute( 'data-wp-on--click' ) ) ) {
+			$p->set_attribute( 'data-wp-on--click', 'core/router::actions.navigate' );
+		}
+		if ( empty( $p->get_attribute( 'data-wp-on--mouseenter' ) ) ) {
+			$p->set_attribute( 'data-wp-on--mouseenter', 'core/router::actions.prefetch' );
```

Comment:
Per @gziolo's comment on https://make.wordpress.org/core/2024/04/09/speculative-loading-in-wordpress/, if we eventually move forward with adopting the Speculation Rules API in WordPress core, we would need to consider whether or how this conflicts with it.

FWIW the Speculation Rules API is a web technology, and if it allows us to replace this manual implementation, that would probably be a good thing. That said, I'm not sure how it would behave here given that we don't do a "real" navigation when those links are clicked but client-side navigation. It might just work out of the box, but that would need to be tested.

I'm mostly commenting for visibility, not something we need to cater for at the moment, as it's going to be a while. But once this has landed in Gutenberg, we may actually want to try using it in combination with the https://wordpress.org/plugins/speculation-rules/ plugin to see what happens.

Reply:
Yeah, speculation rules would most likely need to be disabled on sites with full page client-side navigation. This is because each link click would have a `preventDefault` and so the normal browser navigation would never happen.  So the speculative loads would likely be wasted, especially the `prerender` ones. It may be that `prefetch` speculative loads might be able to be reused, but I'm unsure if the `fetch()` from the Interactivity API could reuse them.

Reply:
Potentially we don't have to completely disable them. There may be ways to configure the rules so that any `a` tags with those specific Interactivity API attributes would be skipped.

This wouldn't be helpful if the entire page is using them, but for pages partially using this it would give the best of both worlds.

Reply:
I just put together a quick test of how Speculation Rules work with client-side navigation: https://speculation-rules-with-client-side-navigation.glitch.me/

In short, `prefetch` _does_ work but `prerender` does not: https://speculation-rules-with-client-side-navigation.glitch.me/

I see that clicking the "with prefetch" link results in `(prefetch cache)` appearing for the entry in the network log, whereas the "with prerender" link is the same as if there was no speculative load at all.

All this to say, it will be important for `prefetch` to be used exclusively on client-side navigation sites. Unfortunately this means they'll miss out on instant page loads that come with prerendering (including loading the subresources like images).

Reply:
Correction: It seems prerendering _will_ add the subresources to the cache, but the prerendered HTML is not served from the prefetch cache for the `fetch()`.

Reply:
Correction 2: It seems Glitch sending `max-age=0` for the speculative prerender response prevented it from being served from cache, while this wasn't the case for the speculative prefetch response. I removed `max-age=0` from that response, and now it seems speculative prerenders _are_ served from cache for client-side navigations via `fetch()`: https://speculation-rules-with-client-side-navigation.glitch.me/

Reply:
I added a follow-up task to [this list](https://github.com/WordPress/gutenberg/pull/59707#issuecomment-2058832203) to explore deeper how both projects interact after this experiment.